http-core/src/main/scala/org/apache/pekko/http/impl/util/StreamUtils.scala

```diff
@@ -396,3 +406,43 @@ private[http] class EnhancedByteStringSource[Mat](val byteStringStream: Source[B
       setHandlers(in, out, this)
     }
 }
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[http] final class SimpleStatefulMap[T, U](
```

Comment:
I think we will need positive and negative tests for this code. It is one thing using a widely used lib function but when you create a new customised equivalent, you need to duplicate a good fraction of the tests from the original lib function.

Reply:
Seems fine, but I will wait @jrudolph for some input before I put more time on this.

Reply:
At the risk of derailing, both this and the inlining discussion really demonstrate the need to have dedicated hardware for benchmarking and a setup so that we can easily compare performance inbetween changes

Reply:
I will write some test this weekend when I back my hometown, wdyt about this approach @jrudolph ?

@mdedetrich Yes, have something like grpc-bench and test the performance continuously would be nice.

Reply:
- [ ]  I have added the test.